File: Plugins/AudioPlayer/src/ios_AudioPlayer/AudioPlayer/AudioBible.swift
//
//  BibleReader.swift
//  AudioPlayer
//
//  Created by Gary Griswold on 7/31/17.
//  Copyright © 2017 ShortSands. All rights reserved.
//

import AVFoundation
import AWS

class AudioBible {
    
    private static var instance: AudioBible?
    static func shared(controller: AudioBibleController) -> AudioBible {
        if (AudioBible.instance == nil) {
            AudioBible.instance = AudioBible(controller: controller)
        }
        return AudioBible.instance!
    }
    
    private let controller: AudioBibleController
    private let controlCenter: AudioControlCenter
    private var audioAnalytics: AudioAnalytics?
    private var player: AVPlayer?
    // Transient Variables
    private var currReference: Reference?
    private var nextReference: Reference?
    
    private init(controller: AudioBibleController) {
        self.controller = controller
        self.controlCenter = AudioControlCenter.shared
    }
    
    deinit {
        print("***** Deinit AudioBible *****")
    }
    
    func getPlayer() -> AVPlayer? {
        return self.player
    }
    
    func getCurrentReference() -> Reference? {
        return self.currReference
    }
    
    func isPlaying() -> Bool {
        return (self.player != nil) ? self.player!.rate > 0.0 : false
    }
    
    func beginReadFile(reference: Reference) {
        print("BibleReader.BEGIN Read File")
        self.currReference = reference
        self.audioAnalytics = AudioAnalytics(mediaSource: "FCBH",
                                             mediaId: reference.damId,
                                             languageId: reference.dpbLanguageCode,
                                             silLang: "User's text lang setting")
        
        print("INSIDE BibleReader \(reference.damId)")
        MediaPlayState.retrieve(mediaId: reference.damId)
        AwsS3Cache.shared.readFile(s3Bucket: reference.getS3Bucket(),
                   s3Key: reference.getS3Key(),
                   expireInterval: Double.infinity,
                   getComplete: {
                    url in
                    if let audioURL = url {
                        self.initAudio(url: audioURL)
                    }
        })
    }
    
    private func initAudio(url: URL) {
        let asset = AVAsset(url: url)
        let playerItem = AVPlayerItem(asset: asset)
        print("Player Item Status \(playerItem.status)")
        
        let seekTime = backupSeek(state: MediaPlayState.currentState)
        if (CMTimeGetSeconds(seekTime) > 0.1) {
            playerItem.seek(to: seekTime)
        }
        self.player = AVPlayer(playerItem: playerItem)
        self.player?.actionAtItemEnd = AVPlayerActionAtItemEnd.none
        self.initNotifications()
        
        self.play()
        self.preFetchNextChapter(reference: self.currReference!)
        
        self.controller.playHasStarted()
        
        self.controlCenter.setupControlCenter(player: self)
        self.controlCenter.nowPlaying(player: self)
    }
    
    private func backupSeek(state: MediaPlayState) -> CMTime {
        if (state.mediaUrl == self.currReference!.toString()) {
            return state.position
        } else {
            return kCMTimeZero
        }
    }
    
    func play() {
        if let play = self.player {
            if let reference = self.currReference {
                print("\n*********** PLAY *************")
                play.play()
                self.audioAnalytics?.playStarted(item: reference.toString(), position: play.currentTime())
            }
        }
    }
    
    func pause() {
        if let play = self.player {
            if let reference = self.currReference {
                print("\n*********** PAUSE *************")
                play.pause()
                self.audioAnalytics?.playEnded(item: reference.toString(), position: play.currentTime())
                self.updateMediaPlayStateTime(reference: reference)
            }
        }
    }
    
    func stop() {
        if self.player != nil && self.player!.rate > 0.0 {
            self.pause()
        }
        //self.removeNotifications()
        self.controller.playHasStopped()
    }
    
    /** This method is called by AudioControlCenter.swift when user clicks the next button. */
    func nextChapter() {
        if let curr = self.nextReference {
            self.currReference = curr
            self.addNextChapter(reference: curr)
            self.preFetchNextChapter(reference: curr)
        } else {
            self.stop()
            MediaPlayState.clear() // Must do after stop, because stop updates
        }
    }
    
    /** This method is called by AudioControlCenter.swift when user clicks the prior button. */
    func priorChapter() {
        if let item = self.player?.currentItem {
            if item.currentTime().seconds < 1.0 {
                if let prior = self.currReference?.priorChapter() {
                    self.nextReference = self.currReference
                    self.currReference = prior
                    self.addNextChapter(reference: self.currReference!)
                }
            } else {
                item.seek(to: kCMTimeZero)
                self.controlCenter.nowPlaying(player: self)
            }
        }
        self.player?.currentItem?.seek(to: kCMTimeZero)
    }
    
    private func initNotifications() {
        let notify = NotificationCenter.default
        notify.addObserver(self,
                           selector: #selector(playerItemDidPlayToEndTime(note:)),
                           name: .AVPlayerItemDidPlayToEndTime,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(playerItemFailedToPlayToEndTime(note:)),
                           name: .AVPlayerItemFailedToPlayToEndTime,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(playerItemPlaybackStalled(note:)),
                           name: .AVPlayerItemPlaybackStalled,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(playerItemNewErrorLogEntry(note:)),
                           name: .AVPlayerItemNewErrorLogEntry,
                           object: nil)
        /*notify.addObserver(self,
                            selector: #selector(playerItemNewAccessLogEntry(note:)),
                            name: .AVPlayerItemNewAccessLogEntry,
                            object: nil)*/
        /*notify.addObserver(self,
                            selector: #selector(playerItemTimeJumped(note:)),
                            name: .AVPlayerItemTimeJumped,
                            object: nil)*/
        notify.addObserver(self,
                           selector: #selector(applicationDidFinishLaunching(note:)),
                           name: .UIApplicationDidFinishLaunching,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(applicationWillEnterForeground(note:)),
                           name: .UIApplicationWillEnterForeground,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(applicationDidEnterBackground(note:)),
                           name: .UIApplicationDidEnterBackground,
                           object: nil)
        notify.addObserver(self,
                           selector: #selector(applicationWillTerminate(note:)),
                           name: .UIApplicationWillTerminate,
                           object: nil)
    }
/*
    private func removeNotifications() {
        print("\n ***** INSIDE REMOVE NOTIFICATIONS")
        let notify = NotificationCenter.default
        notify.removeObserver(self, name: .AVPlayerItemDidPlayToEndTime, object: nil)
        notify.removeObserver(self, name: .AVPlayerItemFailedToPlayToEndTime, object: nil)
        notify.removeObserver(self, name: .AVPlayerItemPlaybackStalled, object: nil)
        notify.removeObserver(self, name: .AVPlayerItemNewErrorLogEntry, object: nil)
        //notify.removeObserver(self, name: .AVPlayerItemNewAccessLogEntry, object: nil)
        //notify.removeObserver(self, name: .AVPlayerItemTimeJumped, object: nil)
 
        notify.removeObserver(self, name: .UIApplicationDidFinishLaunching, object: nil)
        notify.removeObserver(self, name: .UIApplicationWillEnterForeground, object: nil)
        notify.removeObserver(self, name: .UIApplicationDidEnterBackground, object: nil)
        notify.removeObserver(self, name: .UIApplicationWillTerminate, object: nil)
    }
*/
    @objc private func playerItemDidPlayToEndTime(note:Notification) {
        print("\n** DID PLAY TO END \(String(describing: note.object))")
        self.nextChapter()
    }
    @objc private func playerItemFailedToPlayToEndTime(note:Notification) {
        print("\n********* FAILED TO PLAY TO END *********\(String(describing: note.object))")
    }
    @objc private func playerItemPlaybackStalled(note:Notification) {
        print("\n****** PLAYBACK STALLED \(String(describing: note.object))")
    }
    @objc private func playerItemNewErrorLogEntry(note:Notification) {
        print("\n****** ERROR LOG ENTRY \(String(describing: note.object))\n\(String(describing: self.player?.currentItem?.errorLog()))")
    }
    @objc private func playerItemNewAccessLogEntry(note:Notification) {
        print("\n****** ACCESS LOG ENTRY \(String(describing: note.object))\n\(String(describing: self.player?.currentItem?.accessLog()))")
    }
    @objc private func playerItemTimeJumped(note:Notification) {
        print("\n****** TIME JUMPED \(String(describing: note.object))")
    }
    
    /**
    * This method is called when an App is first, launched, not when it is restarted in foreground
    */
    @objc private func applicationDidFinishLaunching(note:Notification) {
        print("\n****** APP DID FINISH LAUNCHING \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
    }
    /**
    * This method is called when an already running App, which went to background, restarts in foreground.
    */
    @objc private func applicationWillEnterForeground(note:Notification) {
        print("\n****** APP WILL ENTER FOREGROUND \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
    }
    /**
    * This method is called when the Home button or Lock button terminate the foreground App
    */
    @objc private func applicationDidEnterBackground(note:Notification) {
        print("\n****** APP DID ENTER BACKGROUND \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
    }
    /**
    * This method is called when the App is fully killed.  It is not affected by the Home or Lock button
    */
    @objc private func applicationWillTerminate(note:Notification) {
        print("\n****** APP WILL TERMINATE \(String(describing: note.object)) \(Date().timeIntervalSince1970)")
        self.stop()
    }
    
    private func updateMediaPlayStateTime(reference: Reference) {
        var result: CMTime = kCMTimeZero
        if let audioChapter = reference.audioChapter {
            if let currentTime = self.player?.currentTime() {
                let verse: Int = audioChapter.findVerseByPosition(priorVerse: 1, time: currentTime)
                result = audioChapter.findPositionOfVerse(verse: verse)
            }
        }
        MediaPlayState.update(url: reference.toString(), time: result)
    }
    
    private func addNextChapter(reference: Reference) {
        self.player?.pause()
        AwsS3Cache.shared.readFile(s3Bucket: reference.getS3Bucket(),
                                   s3Key: reference.getS3Key(),
                                   expireInterval: Double.infinity,
                                   getComplete: {
                                    url in
                                    if let audioURL = url {
                                        let asset = AVAsset(url: audioURL)
                                        let playerItem = AVPlayerItem(asset: asset)
                                        self.player?.replaceCurrentItem(with: playerItem)
                                        self.player?.play()
                                        self.controlCenter.nowPlaying(player: self)
                                    }
        })
    }
    
    private func preFetchNextChapter(reference: Reference) {
        self.readVerseMetaData(reference: reference)
        self.nextReference = reference.nextChapter()
        if let next = self.nextReference {
            AwsS3Cache.shared.readFile(s3Bucket: next.getS3Bucket(),
                                       s3Key: next.getS3Key(),
                                       expireInterval: Double.infinity,
                                       getComplete: { url in })
        }
    }
    
    private func readVerseMetaData(reference: Reference) {
        let reader = MetaDataReader()
        reader.readVerseAudio(damid: reference.damId, sequence: reference.sequence, bookId: reference.book, chapter: reference.chapter, complete: {
            audioChapter in
            if (audioChapter != nil) {
                reference.audioChapter = audioChapter
                //print("PARSED DATA \(self.audioChapter?.toString())")
            }
        })
    }
}

File: YourBible/plugins/com-shortsands-audioplayer/src/android_AudioPlayer/app/src/main/java/com/shortsands/audioplayer/AudioBible.java
package com.shortsands.audioplayer;

import android.media.AudioManager;
import android.media.MediaPlayer;
import android.os.PowerManager;
import android.util.Log;
import com.shortsands.aws.AwsS3Cache;
import com.shortsands.aws.CompletionHandler;
import java.io.File;
import java.io.IOException;

/**
 * Created by garygriswold on 8/30/17.
 */

public class AudioBible implements MediaPlayer.OnErrorListener, MediaPlayer.OnCompletionListener,
        MediaPlayer.OnSeekCompleteListener {

    private static final String TAG = "AudioBible";

    private final AudioBibleController controller;
    private final TOCAudioBible tocAudioBible;
    private final AudioAnalytics audioAnalytics;
    // Transient Variables
    private Reference currReference;
    private Reference nextReference;
    private MediaPlayer mediaPlayer;
    private MediaPlayer nextPlayer;

    public AudioBible(AudioBibleController controller, TOCAudioBible tocBible, Reference reference) {
        super();
        this.controller = controller;
        this.tocAudioBible = tocBible;
        this.audioAnalytics = new AudioAnalytics(controller.activity,
                tocBible.mediaSource,
                reference.damId,
                tocBible.languageCode,
                "User's text lang setting");
        this.currReference = reference;
        this.nextReference = null;
        this.mediaPlayer = null;
        this.nextPlayer = null;
        MediaPlayState.retrieve(this.controller.activity, reference.damId, reference.getS3Key());
    }

    Reference getCurrReference() { return(this.currReference); }

    void beginReadFile() {
        Log.d(TAG, "BibleReader.BEGIN Read File");
        BeginReadFileCompletion handler = new BeginReadFileCompletion();
        AwsS3Cache.shared().readFile(this.currReference.getS3Bucket(),
                this.currReference.getS3Key(),
                Integer.MAX_VALUE,
                handler);
    }

    class BeginReadFileCompletion implements CompletionHandler {
        @Override
        public void completed(Object result) {
            if (result instanceof File) {
                File file = (File) result;
                initAudio(file.getAbsolutePath());
            }
        }
        @Override
        public void failed(Throwable exception) {
            Log.d(TAG, "BeginReadFile Failed " + exception.toString());
        }
    }

    private void initAudio(String url) {
        if (url != null) {
            this.mediaPlayer = this.initPlayer(url);
            if (this.mediaPlayer != null) {
                long seekTime = MediaPlayState.currentState.position;
                if (seekTime > 100L) {
                    this.mediaPlayer.setOnSeekCompleteListener(this);
                    this.mediaPlayer.seekTo((int) seekTime);
                } else {
                    this.onSeekComplete(this.mediaPlayer);
                }
            }
        } else {
            Log.e(TAG, "URL is null");
        }
    }

    @Override
    public void onSeekComplete(MediaPlayer player) {
        player.setOnSeekCompleteListener(null);
        this.play();
        this.preFetchNextChapter(this.currReference);
        this.controller.playHasStarted(player);
    }

    private MediaPlayer initPlayer(String url) {
        MediaPlayer player = new MediaPlayer();
        player.setLooping(false);
        player.setAudioStreamType(AudioManager.STREAM_VOICE_CALL); // Replace with AudioAttributes in SDK 21
        try {
            player.setDataSource(url);
            player.prepare();
            player.setOnCompletionListener(this);
            player.setWakeMode(this.controller.activity, PowerManager.PARTIAL_WAKE_LOCK);
            return player;
        } catch(IOException ioe) {
            Log.e(TAG, "Error in AudioBible.initPlayer " + ioe.toString());
            return null;
        }
    }

    void play() {
        Log.d(TAG, "\n*********** PLAY *************");
        this.mediaPlayer.start();
        long currTime = (this.mediaPlayer != null) ? this.mediaPlayer.getCurrentPosition() : 0;
        this.audioAnalytics.playStarted(this.currReference.toString(), currTime);
    }

    void pause() {
        Log.d(TAG, "\n*********** PAUSE *************");
        this.mediaPlayer.pause();
        long currTime = (this.mediaPlayer != null) ? this.mediaPlayer.getCurrentPosition() : 0;
        this.audioAnalytics.playEnded(this.currReference.toString(), currTime);
        this.updateMediaPlayStateTime();
    }

    void stop() {
        if (this.mediaPlayer.isPlaying()) {
            this.pause();
        }
        this.controller.playHasStopped();
        if (this.mediaPlayer != null) {
            this.mediaPlayer.reset();
            this.mediaPlayer.release();
            this.mediaPlayer = null;
        }
        if (this.nextPlayer != null) {
            this.nextPlayer.reset();
            this.nextPlayer.release();
            this.nextPlayer = null;
        }
    }

    @Override
    public void onCompletion(MediaPlayer player) {
        this.advanceToNextItem();
    }

    @Override
    public boolean onError(MediaPlayer player, int what, int extra) {
        //this.errorTime = System.currentTimeMillis();
        String message;
        switch (what) {
            case MediaPlayer.MEDIA_ERROR_IO:
                message = "MEDIA ERROR IO";
                break;
            case MediaPlayer.MEDIA_ERROR_MALFORMED:
                message = "MEDIA ERROR MALFORMED";
                break;
            case MediaPlayer.MEDIA_ERROR_NOT_VALID_FOR_PROGRESSIVE_PLAYBACK:
                message = "MEDIA ERROR NOT VALID FOR PROGRESSIVE PLAYBACK";
                break;
            case MediaPlayer.MEDIA_ERROR_SERVER_DIED:
                message = "MEDIA ERROR SERVER DIED";
                break;
            case MediaPlayer.MEDIA_ERROR_TIMED_OUT:
                message = "MEDIA ERROR TIMED OUT";
                break;
            case MediaPlayer.MEDIA_ERROR_UNKNOWN:
                message = "MEDIA ERROR UNKNOWN";
                break;
            case MediaPlayer.MEDIA_ERROR_UNSUPPORTED:
                message = "MEDIA ERROR UNSUPPORTED";
                break;
            default:
                message = "Unknown Error " + what;
        }
        Log.e(TAG, "onError " + message + " " + extra);

        mediaPlayer.reset();
        //this.onErrorRecovery = true;
        //this.progressBar.setVisibility(View.VISIBLE);
        //Uri videoUri = Uri.parse(this.videoUrl);
        //this.videoView.setVideoURI(videoUri);

        return true;
    }

    private void updateMediaPlayStateTime() {
        TOCAudioChapter chapter = this.currReference.audioChapter;
        int position = this.mediaPlayer.getCurrentPosition();
        if (chapter != null) {
            int verseNum = chapter.findVerseByPosition(1, position);
            position = chapter.findPositionOfVerse(verseNum);
        } else {
            position -= 3000;
            position = (position >= 0) ? position : 0;
        }
        MediaPlayState.update(this.controller.activity, this.currReference.getS3Key(), (int)position);
    }

    void advanceToNextItem() {
        if (this.nextReference != null) {
            this.currReference = this.nextReference;
            this.addNextChapter(this.currReference);
            this.preFetchNextChapter(this.currReference);
        } else {
            this.stop();
            MediaPlayState.clear(this.controller.activity); // Must be after stop, because stop does update
        }
    }

    private void addNextChapter(Reference reference) {
        if (this.mediaPlayer.isPlaying()) {
            this.mediaPlayer.stop();
            this.mediaPlayer = this.nextPlayer;
            this.mediaPlayer.start();
        } else {
            this.mediaPlayer = this.nextPlayer;
        }
        this.controller.playHasStarted(this.mediaPlayer);
    }

    private void preFetchNextChapter(Reference reference) {
        this.readVerseMetaData(reference);
        this.nextReference = this.tocAudioBible.nextChapter(reference);
        if (this.nextReference != null) {
            PreFetchCompletion handler = new PreFetchCompletion();
            AwsS3Cache.shared().readFile(this.nextReference.getS3Bucket(),
                    this.nextReference.getS3Key(),
                    Integer.MAX_VALUE,
                    handler);
        } else {
            nextPlayer = null;
        }
    }
    class PreFetchCompletion implements CompletionHandler {
        @Override
        public void completed(Object result) {
            if (result instanceof File) {
                File file = (File) result;
                nextPlayer = initPlayer(file.getAbsolutePath());
                mediaPlayer.setNextMediaPlayer(nextPlayer);
            }
        }
        @Override
        public void failed(Throwable exception) {
            Log.d(TAG, "NextReadFile Failed " + exception.toString());
        }
    }

    private void readVerseMetaData(Reference reference) {
        ReadVerseMetaDataHandler handler = new ReadVerseMetaDataHandler(reference);
        MetaDataReader reader = new MetaDataReader(this.controller.activity);
        reader.readVerseAudio(reference.damId, reference.sequence, reference.book, reference.chapter, handler);
    }

    class ReadVerseMetaDataHandler implements CompletionHandler {
        private Reference reference;
        ReadVerseMetaDataHandler(Reference ref) {
            this.reference = ref;
        }
        public void completed(Object result) {
            if (result instanceof TOCAudioChapter) {
                TOCAudioChapter chapter = (TOCAudioChapter)result;
                reference.audioChapter = chapter;
                //Log.d(TAG, "************" + chapter.toString());
            } else {
                this.failed(new Exception("Could not cast to TOCAudioChapter"));
            }
        }
        public void failed(Throwable exception) {
            Log.e(TAG, "Exception in ReadVerseMetaDataHandler " + exception.toString());
        }
    }
}